In [29]:
%pip install openpyxl

     ------------------------------------ 250.9/250.9 kB 570.3 kB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
# Na primeira célula do notebook, sempre se faz as importações das bibliotecas a serem utilizadas.
# Se algum error aparecer, a biblioteca não está instalada.
# Orientação: abrir terminal e rodar: pip install pandas openpyxl

import pandas as pd # Biblioteca de análise de dados
import re           # expressão regulares (para limpar texto)
from datetime import datetime   # Para trabalhar com datas

# Confirmar se as importações funcionaram

print(' pandas versão: ', pd.__version__)
print(' Bibliotecas importadas com sucesso!')

 pandas versão:  3.0.2
 Bibliotecas importadas com sucesso!


In [3]:
# pd.read_csv() >>>> Para ler arquivos .csv
# Caminho - onde está o arquivo. Como ele está ne mesma pasta que este notebook
# basta passar o nome do arquivo, entre aspas, para que a função leia o arquivo e crie um dataframe (tabela) com os dados.

CAMINHO = r"C:\Users\Bete\Desktop\SCTEC_BI\alunos\daniel_roberto\semana_04\base_rh.csv"

df = pd.read_csv(
    CAMINHO,
    sep=';',            # Separador de colunas é ';' e não ','
    encoding='cp1252',  # encoding do Windows, para ler acentos.
    decimal=','         # ',' separador decimal
)

print('Dados carregados com sucesso!')
print(f' Total de linhas : {df.shape[0]}')
print(f' Total de colunas : {df.shape[1]}')

Dados carregados com sucesso!
 Total de linhas : 1000
 Total de colunas : 10


In [10]:
# df.head(n) para ver as primeiras linhas da tabela e confirmar que os dados foram lidos.
# Padrão: df.head(5)
# Use para ter uma rápida visão geral dos dados e confirmar que as colunas foram lidas corretamente e os dados no formato esperado.

df.head(10)


,ID_Funcionario,Nome,Departamento,Cargo,Salario,Data_Admissao,Genero,Idade,Estado_Civil,Status
0,1,Julia Nunes,Logística,Coordenador,9088.34,13/08/2024,F,43,Solteiro,Inativo
1,2,Sr. Gustavo Duarte,TI,Gerente,8155.98,29/04/2017,F,59,Solteiro,Inativo
2,3,Srta. Mariana Cunha,RH,Coordenador,14027.93,11/12/2024,F,27,Divorciado,Ativo
3,4,Ana Sophia da Cruz,Produção,Técnico,6647.97,16/06/2019,M,50,Casado,Inativo
4,5,Dr. Nicolas Pinto,Produção,Gerente,8730.19,29/03/2019,F,62,Casado,Inativo
5,6,Marcela Costa,RH,Analista,12403.42,23/07/2024,M,55,Solteiro,Ativo
6,7,João Miguel da Costa,Produção,Técnico,4771.89,11/11/2021,F,58,Viúvo,Inativo
7,8,Gabrielly Cardoso,RH,Analista,2728.36,26/04/2022,M,56,Divorciado,Ativo
8,9,Raul da Paz,Logística,Coordenador,3164.23,30/11/2022,F,51,Solteiro,Ativo
9,10,Davi Lucca Teixeira,Produção,Gerente,13512.53,29/06/2020,M,35,Viúvo,Ativo


In [8]:
df.tail(10)

,ID_Funcionario,Nome,Departamento,Cargo,Salario,Data_Admissao,Genero,Idade,Estado_Civil,Status
990,991,Sr. Vicente Silveira,Produção,Gerente,8026.70,11/07/2021,F,55,Viúvo,Inativo
991,992,Emanuel da Cruz,Logística,Gerente,12572.11,30/01/2024,M,33,Viúvo,Ativo
992,993,Enrico Monteiro,RH,Coordenador,14677.33,26/12/2017,M,52,Divorciado,Inativo
993,994,Ana Lívia Cardoso,RH,Coordenador,14912.48,22/10/2015,M,27,Solteiro,Inativo
994,995,Fernando Monteiro,RH,Assistente,10983.02,09/01/2023,F,42,Viúvo,Ativo
995,996,Matheus Aragão,Financeiro,Analista,11389.03,06/05/2022,F,26,Viúvo,Ativo
996,997,Ryan Martins,Financeiro,Analista,14060.42,05/06/2023,F,37,Viúvo,Ativo
997,998,Isabelly Ferreira,TI,Gerente,2348.73,29/12/2023,F,58,Solteiro,Inativo
998,999,Thales Farias,Produção,Analista,9430.62,25/12/2019,M,41,Solteiro,Ativo
999,1000,Sra. Maysa Costela,Produção,Técnico,12549.72,14/11/2022,M,46,Viúvo,Inativo


In [9]:
print("Colunas do Dataset: ")
print(df.columns.tolist())

Colunas do Dataset: 
['ID_Funcionario', 'Nome', 'Departamento', 'Cargo', 'Salario', 'Data_Admissao', 'Genero', 'Idade', 'Estado_Civil', 'Status']


## Diagnóstico: tipos e nulos

**O que são os tipos de dados?**
Cada coluna armazena um tipo de informação. O Python precisa saber o tipo para saber o que pode fazer com ela.

| Tipo | O que é | Exemplo |
|---|---|---|
| `int64` | Número Inteiro | 1, 42, 1000 |
| `float64` | Número Decimal | 9088.34, 3.14 |
| `object` | Texto / String | "Julia Nunes", "Ativo" |
| `datetime64` | Data e Hora | 2024-08-13 |

**O que são nulos?**
São células vazias - o campo não foi preenchido no sistema de origem.
Em Python/pandas, nulos aparecem como 'NaN' ("Not a Number") ou 'NaT' ("Not a Time").


In [12]:
# df.info() -> resumo completo do DataFrame em uma só chamada:
# - Nome de cada coluna
# - Quantidade de valores não-nulos
# - Tipo de dado (dtype)
# É o comando mais importante para o primeiro diagnóstico.

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID_Funcionario  1000 non-null   int64  
 1   Nome            1000 non-null   str    
 2   Departamento    1000 non-null   str    
 3   Cargo           1000 non-null   str    
 4   Salario         1000 non-null   float64
 5   Data_Admissao   1000 non-null   str    
 6   Genero          1000 non-null   str    
 7   Idade           1000 non-null   int64  
 8   Estado_Civil    1000 non-null   str    
 9   Status          1000 non-null   str    
dtypes: float64(1), int64(2), str(7)
memory usage: 78.3 KB


In [15]:
# df.isnull() > cria uma tabela de True/False (True =  é nulo)
# .sum() > soma todos os True de cada coluna (True vale 1, False vale 0)
# Resultado: quantidade de nulos por coluna.

print("Quantidade de nulos por coluna: ")
print(df.isnull().sum())

print()

# Percentual de nulos - mais útil em bases grandes, para entender a proporçãod e nulos em  cada coluna.
# len(df) > quantidade total de linhas do DataFrame.

print("Percentual de nulos por coluna: ")
print((df.isnull().sum() / len(df) * 100).round(2))

Quantidade de nulos por coluna: 
ID_Funcionario    0
Nome              0
Departamento      0
Cargo             0
Salario           0
Data_Admissao     0
Genero            0
Idade             0
Estado_Civil      0
Status            0
dtype: int64

Percentual de nulos por coluna: 
ID_Funcionario    0.0
Nome              0.0
Departamento      0.0
Cargo             0.0
Salario           0.0
Data_Admissao     0.0
Genero            0.0
Idade             0.0
Estado_Civil      0.0
Status            0.0
dtype: float64


---
## Estatísticas das colunas numéricas

`describe()` gera um resumo estatístico automático.  
Você vai usar isso para detectar valores suspeitos (ex: salário de R$0,00 ou idade de 200 anos).

| Estatística | O que significa |
|---|---|
| `count` | Quantos valores não-nulos existem |
| `mean` | Média |
| `std` | Desvio padrão — o quanto os valores variam em torno da média |
| `min` | Menor valor |
| `25%` | 25% dos registros estão abaixo desse valor |
| `50%` | Metade dos registros estão abaixo (mediana) |
| `75%` | 75% dos registros estão abaixo desse valor |
| `max` | Maior valor |

In [16]:
# df.describe() → estatísticas das colunas numéricas
# .round(2)    → arredonda para 2 casas decimais

df.describe().round(2)

,ID_Funcionario,Salario,Idade
count,1000.00,1000.00,1000.00
mean,500.50,8579.95,41.40
std,288.82,3657.37,13.67
min,1.00,2000.71,18.00
25%,250.75,5564.54,30.00
50%,500.50,8571.13,41.00
75%,750.25,11554.63,53.00
max,1000.00,14954.51,65.00


---
## Explorando colunas categóricas

**Coluna categórica** = coluna com poucos valores possíveis (departamento, cargo, status...)  

`nunique()` conta quantos valores diferentes existem.  
`value_counts()` mostra a distribuição — quantas vezes cada valor aparece.

In [17]:
# df.nunique() → retorna quantos valores ÚNICOS cada coluna tem
# Serve para identificar quais são categóricas (poucas opções)
# e quais são identificadoras (muitas opções diferentes)

print('Valores únicos por coluna:')
print(df.nunique())

Valores únicos por coluna:
ID_Funcionario    1000
Nome               976
Departamento         6
Cargo                5
Salario           1000
Data_Admissao      878
Genero               2
Idade               48
Estado_Civil         4
Status               2
dtype: int64


In [18]:
# df['coluna'].value_counts() → conta quantas vezes cada valor aparece,
# ordenado do mais frequente para o menos frequente

# Vamos ver todas as colunas categóricas de uma vez usando um loop:
# 'for col in lista' = para cada coluna na lista, executa o bloco abaixo

for col in ['Departamento', 'Cargo', 'Genero', 'Status', 'Estado_Civil']:
    print(f'\n── {col} ──')   # f-string: {} insere o valor da variável
    print(df[col].value_counts())


── Departamento ──
Departamento
Financeiro    189
Produção      182
RH            166
Vendas        160
Logística     156
TI            147
Name: count, dtype: int64

── Cargo ──
Cargo
Coordenador    225
Gerente        221
Assistente     199
Analista       178
Técnico        177
Name: count, dtype: int64

── Genero ──
Genero
F    510
M    490
Name: count, dtype: int64

── Status ──
Status
Ativo      515
Inativo    485
Name: count, dtype: int64

── Estado_Civil ──
Estado_Civil
Solteiro      265
Viúvo         261
Divorciado    240
Casado        234
Name: count, dtype: int64


---
## Selecionando colunas e filtrando linhas

São as operações mais usadas no dia a dia. Você vai usar isso em todo projeto.

**Seleção de coluna** → `df['Nome']`  
**Seleção de múltiplas colunas** → `df[['Nome', 'Salario']]`  
**Filtro de linhas** → `df[df['Status'] == 'Ativo']`

In [19]:
# Selecionar UMA coluna → retorna uma Series (lista com índice)
# Uma Series é como uma coluna de planilha isolada

df['Nome'].head(5)

0            Julia Nunes
1     Sr. Gustavo Duarte
2    Srta. Mariana Cunha
3     Ana Sophia da Cruz
4      Dr. Nicolas Pinto
Name: Nome, dtype: str

In [20]:
# Selecionar MÚLTIPLAS colunas → retorna um DataFrame
# Note os dois colchetes: df[  ['col1', 'col2']  ]
# O primeiro [] é o operador de acesso do DataFrame
# O segundo [] é a lista de colunas

df[['Nome', 'Departamento', 'Cargo', 'Salario']].head(6)

,Nome,Departamento,Cargo,Salario
0,Julia Nunes,Logística,Coordenador,9088.34
1,Sr. Gustavo Duarte,TI,Gerente,8155.98
2,Srta. Mariana Cunha,RH,Coordenador,14027.93
3,Ana Sophia da Cruz,Produção,Técnico,6647.97
4,Dr. Nicolas Pinto,Produção,Gerente,8730.19
5,Marcela Costa,RH,Analista,12403.42


In [21]:
# FILTRO SIMPLES: selecionar só as linhas onde Status == 'Ativo'
# df['Status'] == 'Ativo' → cria uma Series de True/False
# df[ True/False ]        → mantém só as linhas onde é True

ativos = df[df['Status'] == 'Ativo']

print(f'Total de funcionários : {len(df)}')
print(f'Funcionários ativos   : {len(ativos)}')
print(f'Funcionários inativos : {len(df) - len(ativos)}')

Total de funcionários : 1000
Funcionários ativos   : 515
Funcionários inativos : 485


In [22]:
# FILTRO COMPOSTO: duas condições ao mesmo tempo
# Use & para AND (as duas condições precisam ser verdadeiras)
# Use | para OR  (pelo menos uma precisa ser verdadeira)
# IMPORTANTE: cada condição precisa estar entre parênteses

gerentes_ti = df[
    (df['Cargo'] == 'Gerente') & (df['Departamento'] == 'TI')
]

print(f'Gerentes na área de TI: {len(gerentes_ti)}')
print()
gerentes_ti[['Nome', 'Salario', 'Status', 'Data_Admissao']]

Gerentes na área de TI: 31



,Nome,Salario,Status,Data_Admissao
1,Sr. Gustavo Duarte,8155.98,Inativo,29/04/2017
49,Benjamin da Costa,9766.01,Inativo,27/04/2024
130,Catarina Costela,9597.83,Inativo,03/07/2023
131,Laura Cardoso,7364.46,Inativo,29/06/2018
175,Maria Cardoso,12127.27,Inativo,22/05/2024
187,Ana Júlia Cardoso,8738.63,Ativo,30/05/2021
247,Ian Pereira,2395.32,Ativo,10/05/2019
273,Ana Clara Peixoto,7301.14,Ativo,16/11/2019
296,Augusto Santos,8237.80,Inativo,26/08/2021
307,Manuela Barbosa,7015.91,Ativo,21/06/2017


In [23]:
# Estatísticas rápidas de uma coluna numérica
# .mean()  → média
# .max()   → maior valor
# .min()   → menor valor
# .median()→ mediana (valor do meio quando ordenados)

print(f"Salário médio : R$ {df['Salario'].mean():,.2f}")
print(f"Salário máximo: R$ {df['Salario'].max():,.2f}")
print(f"Salário mínimo: R$ {df['Salario'].min():,.2f}")
print(f"Salário mediana:R$ {df['Salario'].median():,.2f}")

# :,.2f → formata o número com separador de milhar (,) e 2 casas decimais (.2f)

Salário médio : R$ 8,579.95
Salário máximo: R$ 14,954.51
Salário mínimo: R$ 2,000.71
Salário mediana:R$ 8,571.13


---
## Exportando para JSON

**O que é JSON?**  
É o formato padrão de troca de dados entre sistemas modernos — APIs, ERPs na nuvem, dashboards web.  
JSON = *JavaScript Object Notation*

```json
[
  { "Nome": "Julia Nunes", "Departamento": "Logística", "Salario": 9088.34 },
  { "Nome": "Gustavo Duarte", "Departamento": "TI", "Salario": 8155.98 }
]
```

**Por que `orient="records"`?**  
Define o formato do JSON. `records` = cada linha vira um objeto separado.  
É o formato que APIs REST esperam receber.

In [24]:
import json  # biblioteca nativa do Python para manipular JSON

# df.to_json() → converte o DataFrame para JSON e salva no arquivo
df.to_json(
    'base_rh.json',
    orient='records',      # cada linha vira um objeto {}
    force_ascii=False,     # False = salva ã, é, ç como estão
    indent=2               # 2 espaços de indentação → arquivo legível
)

print('✓ base_rh.json gerado!')

# Ler de volta para confirmar
df_json = pd.read_json('base_rh.json')
print(f'  Linhas lidas do JSON: {len(df_json)}')

# Visualizar os 2 primeiros registros no formato JSON
with open('base_rh.json', 'r', encoding='utf-8') as f:
    amostra = json.load(f)
print('\nPrimeiros 2 registros no JSON:')
print(json.dumps(amostra[:4], ensure_ascii=False, indent=4))

✓ base_rh.json gerado!
  Linhas lidas do JSON: 1000

Primeiros 2 registros no JSON:
[
    {
        "ID_Funcionario": 1,
        "Nome": "Julia Nunes",
        "Departamento": "Logística",
        "Cargo": "Coordenador",
        "Salario": 9088.34,
        "Data_Admissao": "13/08/2024",
        "Genero": "F",
        "Idade": 43,
        "Estado_Civil": "Solteiro",
        "Status": "Inativo"
    },
    {
        "ID_Funcionario": 2,
        "Nome": "Sr. Gustavo Duarte",
        "Departamento": "TI",
        "Cargo": "Gerente",
        "Salario": 8155.98,
        "Data_Admissao": "29/04/2017",
        "Genero": "F",
        "Idade": 59,
        "Estado_Civil": "Solteiro",
        "Status": "Inativo"
    },
    {
        "ID_Funcionario": 3,
        "Nome": "Srta. Mariana Cunha",
        "Departamento": "RH",
        "Cargo": "Coordenador",
        "Salario": 14027.93,
        "Data_Admissao": "11/12/2024",
        "Genero": "F",
        "Idade": 27,
        "Estado_Civil": "Divorciado"

---
## Exportando para Excel

Excel ainda é o formato mais pedido por gestores em ambientes industriais.  
O pandas permite criar um arquivo com **múltiplas abas** — uma por departamento.

**O que é `ExcelWriter`?**  
É um gerenciador que permite escrever várias abas no mesmo arquivo `.xlsx`.  
O `with ... as writer:` garante que o arquivo seja fechado corretamente ao final.

> ⚠️ Precisa do `openpyxl` instalado: `pip install openpyxl`

In [30]:
# Exportação simples: tudo em uma aba
df.to_excel('base_rh.xlsx', index=False, sheet_name='Funcionarios')
print('✓ base_rh.xlsx (aba única) gerado!')
# Exportação avançada: uma aba por departamento
with pd.ExcelWriter('base_rh_por_depto.xlsx', engine='openpyxl') as writer:

    # Aba 1: dataset completo
    df.to_excel(writer, sheet_name='Todos', index=False)
    print('  Aba Todos: todos os registros')

    # df['Departamento'].unique() → lista de departamentos sem repetição
    # sorted() → ordena em ordem alfabética
    for depto in sorted(df['Departamento'].unique()):

        # Filtrar só os funcionários desse departamento
        df_depto = df[df['Departamento'] == depto]

        # Limite do Excel: nomes de aba têm no máximo 31 caracteres
        nome_aba = depto[:31]

        df_depto.to_excel(writer, sheet_name=nome_aba, index=False)
        print(f'  Aba "{nome_aba}": {len(df_depto)} registros')
print('\n✓ base_rh_por_depto.xlsx gerado!')

✓ base_rh.xlsx (aba única) gerado!
  Aba Todos: todos os registros
  Aba "Financeiro": 189 registros
  Aba "Logística": 156 registros
  Aba "Produção": 182 registros
  Aba "RH": 166 registros
  Aba "TI": 147 registros
  Aba "Vendas": 160 registros

✓ base_rh_por_depto.xlsx gerado!


---
## Módulo datetime: o básico

**Por que isso importa?**  
A coluna `Data_Admissao` veio como **texto** (`object`).  
O Python não sabe que `'13/08/2024'` é uma data — pra ele é só uma string.  
Enquanto for texto, você **não consegue** calcular diferenças, filtrar por período, extrair mês ou ano.

**Dois comandos essenciais:**

| Comando | Significado | Exemplo |
|---|---|---|
| `strptime` | string → parse → time (lê um texto e vira data) | `'13/08/2024'` → `datetime(2024, 8, 13)` |
| `strftime` | string → format → time (pega uma data e formata como texto) | `datetime(2024, 8, 13)` → `'Agosto de 2024'` |

**Códigos de formato de data:**
| Código | Significado | Exemplo |
|---|---|---|
| `%d` | Dia com zero à esquerda | 08 |
| `%m` | Mês com zero à esquerda | 08 |
| `%Y` | Ano com 4 dígitos | 2024 |
| `%B` | Nome do mês por extenso | August |

In [31]:
from datetime import datetime, timedelta

# ── strptime: texto → objeto data ─────────────────────────────────────────
# Pegamos o primeiro valor da coluna Data_Admissao como exemplo
data_texto = '13/08/2024'

# datetime.strptime(texto, formato)
# O formato descreve COMO a string está escrita:
# '%d/%m/%Y' = dia/mês/ano
data_obj = datetime.strptime(data_texto, '%d/%m/%Y')

print(f'String original  : "{data_texto}"')
print(f'Objeto datetime  : {data_obj}')
print(f'Tipo             : {type(data_obj)}')
print()

# ── strftime: objeto data → texto formatado ───────────────────────────────
# Agora podemos formatar a data do jeito que quisermos
formatado = data_obj.strftime('%d de %B de %Y')  # 13 de August de 2024
print(f'Formatado        : {formatado}')

# Extraindo partes da data
print(f'Ano              : {data_obj.year}')
print(f'Mês              : {data_obj.month}')
print(f'Dia              : {data_obj.day}')

String original  : "13/08/2024"
Objeto datetime  : 2024-08-13 00:00:00
Tipo             : <class 'datetime.datetime'>

Formatado        : 13 de August de 2024
Ano              : 2024
Mês              : 8
Dia              : 13


**O que é `timedelta`?**  
É o resultado da subtração entre duas datas — a diferença de tempo.  
Se você subtrair uma data de outra, recebe um `timedelta` com a diferença em dias.

In [32]:
# ── timedelta: diferença entre duas datas ────────────────────────────────
# datetime.today() → data e hora de agora
hoje = datetime.today()

# Subtrair duas datas retorna um objeto timedelta
diferenca = hoje - data_obj

print(f'Data de admissão : {data_texto}')
print(f'Hoje             : {hoje.strftime("%d/%m/%Y")}')
print(f'Diferença        : {diferenca.days} dias')
print(f'Em anos          : {diferenca.days / 365.25:.1f} anos')

# Por que 365.25 e não 365?
# Todo 4 anos existe um ano bissexto com 366 dias.
# 365.25 é a média correta: (365 × 3 + 366) / 4 = 365.25



Data de admissão : 13/08/2024
Hoje             : 09/05/2026
Diferença        : 634 dias
Em anos          : 1.7 anos


---
## Convertendo a coluna inteira com pd.to_datetime

A célula anterior mostrou como converter **uma** string em data.  
Mas nosso DataFrame tem 1.000 linhas.  

`pd.to_datetime()` converte a coluna **inteira de uma vez** — sem loop, muito mais eficiente.

In [33]:
# Antes da conversão:
print(f'Tipo ANTES : {df["Data_Admissao"].dtype}')
print(f'Valor antes: "{df["Data_Admissao"].iloc[0]}"')
# .iloc[0] → acessa a primeira linha pelo índice (iloc = integer location)

print()

# Converter a coluna inteira
# Trabalhamos em uma cópia para não alterar o df original
df_limpo = df.copy()
# .copy() → cria uma cópia independente
# Sem isso, mudanças em df_limpo afetariam o df original

df_limpo['Data_Admissao'] = pd.to_datetime(
    df_limpo['Data_Admissao'],
    format='%d/%m/%Y',  # formato da string de entrada
    errors='coerce'     # datas inválidas viram NaT (Not a Time) em vez de erro
)

print(f'Tipo DEPOIS: {df_limpo["Data_Admissao"].dtype}')
print(f'Valor depois: {df_limpo["Data_Admissao"].iloc[0]}')
print(f'NaT gerados : {df_limpo["Data_Admissao"].isna().sum()}')

Tipo ANTES : str
Valor antes: "13/08/2024"

Tipo DEPOIS: datetime64[us]
Valor depois: 2024-08-13 00:00:00
NaT gerados : 0


---
## Extraindo informações da data com o acessor `.dt`

Depois de converter para datetime, o pandas libera o **acessor `.dt`**  
que dá acesso a propriedades da data:

| Código | O que retorna | Exemplo |
|---|---|---|
| `.dt.year` | Ano | 2024 |
| `.dt.month` | Mês (número) | 8 |
| `.dt.quarter` | Trimestre | 3 |
| `.dt.day_name()` | Nome do dia da semana | "Tuesday" |

In [34]:
# Criar colunas derivadas de data
df_limpo['Ano_Admissao'] = df_limpo['Data_Admissao'].dt.year
df_limpo['Mes_Admissao'] = df_limpo['Data_Admissao'].dt.month
df_limpo['Trimestre']    = df_limpo['Data_Admissao'].dt.quarter

# Calcular tempo de casa em anos
# pd.Timestamp.today() → data atual no formato do pandas
hoje_ts = pd.Timestamp.today()

df_limpo['Tempo_Casa_Anos'] = (
    (hoje_ts - df_limpo['Data_Admissao']).dt.days / 365.25
).round(1)
# Subtração de datas → timedelta
# .dt.days → extrai a diferença em dias
# / 365.25 → converte dias em anos
# .round(1) → arredonda para 1 casa decimal

# Ver o resultado
cols = ['Nome', 'Data_Admissao', 'Ano_Admissao', 'Mes_Admissao', 'Trimestre', 'Tempo_Casa_Anos']
df_limpo[cols].head(6)

,Nome,Data_Admissao,Ano_Admissao,Mes_Admissao,Trimestre,Tempo_Casa_Anos
0,Julia Nunes,2024-08-13,2024,8,3,1.7
1,Sr. Gustavo Duarte,2017-04-29,2017,4,2,9.0
2,Srta. Mariana Cunha,2024-12-11,2024,12,4,1.4
3,Ana Sophia da Cruz,2019-06-16,2019,6,2,6.9
4,Dr. Nicolas Pinto,2019-03-29,2019,3,1,7.1
5,Marcela Costa,2024-07-23,2024,7,3,1.8


---
## Criando sua primeira função

**O que é uma função?**  
É um bloco de código com nome próprio que você pode reutilizar.  
Em vez de copiar o mesmo código 10 vezes, você define uma vez e chama quando precisar.

```python
def nome_da_funcao(parametro):
    # o que a função faz
    return resultado
```

- `def` → palavra-chave que define uma função
- `parametro` → valor de entrada que a função recebe
- `return` → valor de saída que a função entrega
- A indentação (4 espaços) define o que está dentro da função

In [35]:
# Função que classifica o tempo de casa em uma categoria de texto
# Por que criar isso? Categorias são mais fáceis de usar em gráficos
# e filtros no Power BI do que números decimais.

def classificar_tempo(anos):
    """
    Recebe: anos (número) — o tempo de casa do funcionário
    Retorna: uma categoria em texto
    """
    # pd.isna() verifica se o valor é nulo (NaN ou NaT)
    if pd.isna(anos):
        return 'Sem data'
    elif anos < 1:
        return 'Menos de 1 ano'
    elif anos < 3:
        return '1 a 3 anos'
    elif anos < 5:
        return '3 a 5 anos'
    else:
        return 'Mais de 5 anos'


# Testar a função com valores individuais antes de aplicar no DataFrame
print('Testando a função:')
print(f'  0.5 anos → "{classificar_tempo(0.5)}"')
print(f'  2.1 anos → "{classificar_tempo(2.1)}"')
print(f'  4.8 anos → "{classificar_tempo(4.8)}"')
print(f'  9.0 anos → "{classificar_tempo(9.0)}"')

Testando a função:
  0.5 anos → "Menos de 1 ano"
  2.1 anos → "1 a 3 anos"
  4.8 anos → "3 a 5 anos"
  9.0 anos → "Mais de 5 anos"


In [36]:
# .apply(função) → aplica a função em cada linha da coluna
# É como fazer um loop em cada célula, mas de forma mais eficiente

df_limpo['Faixa_Tempo_Casa'] = df_limpo['Tempo_Casa_Anos'].apply(classificar_tempo)

print('Distribuição por faixa de tempo de casa:')
print(df_limpo['Faixa_Tempo_Casa'].value_counts())

Distribuição por faixa de tempo de casa:
Faixa_Tempo_Casa
Mais de 5 anos    537
3 a 5 anos        217
1 a 3 anos        213
Menos de 1 ano     33
Name: count, dtype: int64


---
## Filtrando por período de datas

Agora que a data é um tipo real, podemos fazer filtros por período.  
`pd.DateOffset(years=1)` → um deslocamento de exatamente 1 ano no calendário.  
É mais preciso que `365 dias` porque considera meses com 28/30/31 dias e anos bissextos.

In [37]:
hoje = pd.Timestamp.today()

# Definir os pontos de corte
um_ano_atras    = hoje - pd.DateOffset(years=1)
cinco_anos_atras = hoje - pd.DateOffset(years=5)

# Filtrar por período
# >= significa 'maior ou igual a' → admitidos depois de 1 ano atrás
novos     = df_limpo[df_limpo['Data_Admissao'] >= um_ano_atras]
veteranos = df_limpo[df_limpo['Data_Admissao'] <  cinco_anos_atras]

print(f'Admitidos nos últimos 12 meses : {len(novos)} funcionários')
print(f'Veteranos (5+ anos de casa)    : {len(veteranos)} funcionários')

print()

# Comparativo de salário: novos vs veteranos
print('Salário médio:')
print(f'  Novos     : R$ {novos["Salario"].mean():,.2f}')
print(f'  Veteranos : R$ {veteranos["Salario"].mean():,.2f}')

Admitidos nos últimos 12 meses : 38 funcionários
Veteranos (5+ anos de casa)    : 533 funcionários

Salário médio:
  Novos     : R$ 8,914.83
  Veteranos : R$ 8,600.15


---
## Agrupamentos com groupby

**O que é `groupby`?**  
É como a tabela dinâmica do Excel — agrupa as linhas por uma coluna  
e aplica um cálculo (média, soma, contagem...) em cada grupo.

```
df.groupby('coluna_agrupamento')['coluna_calculo'].função()
```

In [38]:
# Salário médio por departamento
# .mean()             → calcula a média de cada grupo
# .round(2)           → arredonda para 2 casas
# .sort_values(...)   → ordena do maior para o menor
# ascending=False     → decrescente (False = do maior para o menor)

print('Salário médio por departamento:')
print(
    df_limpo.groupby('Departamento')['Salario']
    .mean()
    .round(2)
    .sort_values(ascending=False)
)

Salário médio por departamento:
Departamento
Produção      8968.73
Logística     8881.66
RH            8791.58
Financeiro    8333.12
Vendas        8317.08
TI            8142.91
Name: Salario, dtype: float64


In [39]:
# Tempo médio de casa por departamento
media_depto = (
    df_limpo
    .groupby('Departamento')['Tempo_Casa_Anos']
    .mean()
    .round(1)
    .sort_values(ascending=False)
    .rename('Media_Anos')   # .rename() → renomeia a coluna resultante
    .reset_index()          # .reset_index() → converte o índice em coluna comum
)

print('Tempo médio de casa por departamento:')
print(media_depto.to_string(index=False))

# .idxmax() → retorna o NOME do índice com maior valor
maior = media_depto.set_index('Departamento')['Media_Anos'].idxmax()
menor = media_depto.set_index('Departamento')['Media_Anos'].idxmin()
print(f'\n→ Maior retenção: {maior}')
print(f'→ Menor retenção: {menor}')

Tempo médio de casa por departamento:
Departamento  Media_Anos
          RH         5.8
      Vendas         5.7
    Produção         5.5
          TI         5.5
   Logística         5.3
  Financeiro         5.1

→ Maior retenção: RH
→ Menor retenção: Financeiro


In [40]:
# Admissões por ano — histórico completo
# .size() → conta o número de linhas em cada grupo
# .sort_index() → ordena pelo índice (neste caso, o ano)

por_ano = (
    df_limpo
    .groupby('Ano_Admissao')
    .size()
    .rename('Admissoes')
    .reset_index()
    .dropna()  # remove linhas com Ano_Admissao nulo
)

print('Admissões por ano:')
for _, row in por_ano.iterrows():
    # .iterrows() → itera linha por linha, retornando (índice, Series)
    # _ → descartamos o índice (convenção: _ = 'não preciso desse valor')
    barra = '█' * int(row['Admissoes'] / 5)
    print(f'  {int(row["Ano_Admissao"])}  {row["Admissoes"]:>4}  {barra}')

Admissões por ano:
  2015    33  ██████
  2016    87  █████████████████
  2017    99  ███████████████████
  2018    89  █████████████████
  2019   101  ████████████████████
  2020    94  ██████████████████
  2021    94  ██████████████████
  2022   108  █████████████████████
  2023   106  █████████████████████
  2024   113  ██████████████████████
  2025    76  ███████████████


---
## Salvando o dataset limpo e enriquecido

Aqui fazemos o **Load** do ETL: salvamos o resultado final  
em um arquivo CSV que vai alimentar o Power BI na próxima semana.

In [41]:
# Salvar o DataFrame enriquecido
# encoding='utf-8-sig' → UTF-8 com BOM — necessário para o Excel no Windows
#                         abrir o arquivo sem distorcer ã, é, ç

df_limpo.to_csv(
    'base_rh_dia01.csv',
    index=False,         # não salva o índice numérico como coluna extra
    sep=';',             # separador ponto-e-vírgula
    encoding='utf-8-sig'
)

print('✓ base_rh_dia01.csv salvo!')
print(f'  Colunas finais: {df_limpo.columns.tolist()}')
print(f'  Linhas        : {len(df_limpo)}')

✓ base_rh_dia01.csv salvo!
  Colunas finais: ['ID_Funcionario', 'Nome', 'Departamento', 'Cargo', 'Salario', 'Data_Admissao', 'Genero', 'Idade', 'Estado_Civil', 'Status', 'Ano_Admissao', 'Mes_Admissao', 'Trimestre', 'Tempo_Casa_Anos', 'Faixa_Tempo_Casa']
  Linhas        : 1000


---
## Exercício final

Coloque em prática tudo que você aprendeu hoje.  
Resolva os passos abaixo **nesta célula** (pode criar mais células se quiser).

---

### Cenário
O gestor de RH da Indústria Metalúrgica Catarinense pediu 2 entregáveis:
1. A base de funcionários em Excel com **uma aba por departamento**
2. Um relatório de tempo de casa pronto para o Power BI

---

### Passos

**1.** Carregue o `base_rh.csv` com os parâmetros corretos e confirme que `Salario` é `float64`.

**2.** Use `df.info()` e `df.describe()` para fazer o diagnóstico. Anote em uma célula Markdown:  
→ *"Os 3 problemas mais importantes que encontrei foram..."*

**3.** Converta `Data_Admissao` para datetime. Crie as colunas:  
`Ano_Admissao`, `Mes_Admissao`, `Tempo_Casa_Anos` e `Faixa_Tempo_Casa`.

**4.** Responda com `groupby`:  
→ Qual departamento tem a **maior** média de tempo de casa?  
→ Qual tem a **menor**?

**5.** Exporte `base_rh_deptos.xlsx` com uma aba `'Completo'` e mais uma aba por departamento.

**6.** Faça o commit:  
```bash
git pull origin master
git add alunos/seunome/
git commit -m "semana 04 - dia 01: entrega arquivos e datetime"
git push origin master
git log --oneline -3
```
Cole o resultado do `git log` na última célula do notebook.

---

> 💡 **Dica para o Passo 4:**  
> `df.groupby('Departamento')['Tempo_Casa_Anos'].mean().sort_values(ascending=False)`

> ⚠️ **Se o push falhar com 'rejected':** você esqueceu o `git pull` antes.

In [1]:
# ── Resolva o exercício aqui ──────────────────────────────────────────────
# Passo 1: carregue o CSV
import pandas as pd
from datetime import datetime

df = pd.read_csv("base_rh.csv", sep=";", encoding="cp1252", decimal=",")

# Passo 2: diagnóstico
print(df.info())
print(df.describe())

# Passo 3: converter datas e criar colunas derivadas
df["Data_Admissao"] = pd.to_datetime(df["Data_Admissao"], dayfirst=True)
df["Ano_Admissao"] = df["Data_Admissao"].dt.year
df["Mes_Admissao"] = df["Data_Admissao"].dt.month

# Passo 4: groupby — maior e menor média de tempo de casa
hoje = datetime.now()
df['Tempo_Casa_Anos'] = (hoje - df['Data_Admissao']).dt.days / 365.25

# Criando Faixas (Exemplo de lógica para o Power BI)
df['Faixa_Tempo_Casa'] = pd.cut(df['Tempo_Casa_Anos'], 
                                bins=[0, 2, 5, 10, 100], 
                                labels=['Até 2 anos', '2-5 anos', '5-10 anos', 'Mais de 10 anos'])

# Ranking de Tempo de Casa
ranking_tempo = df.groupby('Departamento')['Tempo_Casa_Anos'].mean().sort_values(ascending=False)
print(f"Maior média de tempo: {ranking_tempo.idxmax()} ({ranking_tempo.max():.1f} anos)")
print(f"Menor média de tempo: {ranking_tempo.idxmin()} ({ranking_tempo.min():.1f} anos)")

# Passo 5: exportar para Excel
with pd.ExcelWriter('base_rh_por_dpto.xlsx', engine='openpyxl') as writer:
    # Aba principal
    df.to_excel(writer, sheet_name='Completo', index=False)

    # Abas por departamento
    for depto in sorted(df['Departamento'].unique()):
        df_depto = df[df['Departamento'] == depto]
        df_depto.to_excel(writer, sheet_name=depto[:31], index=False)

# Passo 6: cole o output do git log aqui como comentário
# git log:

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID_Funcionario  1000 non-null   int64  
 1   Nome            1000 non-null   str    
 2   Departamento    1000 non-null   str    
 3   Cargo           1000 non-null   str    
 4   Salario         1000 non-null   float64
 5   Data_Admissao   1000 non-null   str    
 6   Genero          1000 non-null   str    
 7   Idade           1000 non-null   int64  
 8   Estado_Civil    1000 non-null   str    
 9   Status          1000 non-null   str    
dtypes: float64(1), int64(2), str(7)
memory usage: 78.3 KB
None
       ID_Funcionario       Salario        Idade
count     1000.000000   1000.000000  1000.000000
mean       500.500000   8579.951780    41.399000
std        288.819436   3657.365307    13.665817
min          1.000000   2000.710000    18.000000
25%        250.750000   5564.545000    30.000000
50%    